packages/camera/android/src/main/java/io/flutter/plugins/camera/CameraPlugin.java

```diff
@@ -4,172 +4,107 @@
 
 package io.flutter.plugins.camera;
 
-import android.hardware.camera2.CameraAccessException;
+import android.app.Activity;
 import android.os.Build;
+import android.util.Log;
 import androidx.annotation.NonNull;
-import io.flutter.plugin.common.EventChannel;
-import io.flutter.plugin.common.MethodCall;
-import io.flutter.plugin.common.MethodChannel;
-import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
-import io.flutter.plugin.common.MethodChannel.Result;
+import androidx.annotation.Nullable;
+import io.flutter.embedding.engine.plugins.FlutterPlugin;
+import io.flutter.embedding.engine.plugins.activity.ActivityAware;
+import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
+import io.flutter.plugin.common.BinaryMessenger;
 import io.flutter.plugin.common.PluginRegistry.Registrar;
-import io.flutter.view.FlutterView;
+import io.flutter.plugins.camera.CameraPermissions.PermissionsRegistry;
 import io.flutter.view.TextureRegistry;
 
-public class CameraPlugin implements MethodCallHandler {
+/**
+ * Plugin implementation that uses the {@code io.flutter.embedding} package.
+ *
+ * <p>Instantiate this in an add to app scenario to gracefully handle activity and context changes.
+ */
+public final class CameraPlugin implements FlutterPlugin, ActivityAware {
 
-  private final CameraPermissions cameraPermissions = new CameraPermissions();
-  private final FlutterView view;
-  private final Registrar registrar;
-  private final EventChannel imageStreamChannel;
-  private Camera camera;
+  private static final String TAG = "CameraPlugin";
+  private @Nullable FlutterPluginBinding flutterPluginBinding;
+  private @Nullable MethodCallHandlerImpl methodCallHandler;
 
-  private CameraPlugin(Registrar registrar) {
-    this.registrar = registrar;
-    this.view = registrar.view();
-    this.imageStreamChannel =
-        new EventChannel(registrar.messenger(), "plugins.flutter.io/camera/imageStream");
-  }
+  /**
+   * Initialize this within the {@code #configureFlutterEngine} of a Flutter activity or fragment.
+   *
+   * <p>See {@code io.flutter.plugins.camera.MainActivity} for an example.
+   */
+  public CameraPlugin() {}
 
+  /**
+   * Registers a plugin implementation that uses the stable {@code io.flutter.plugin.common}
+   * package.
+   *
+   * <p>Calling this automatically initializes the plugin. However plugins initialized this way
+   * won't react to changes in activity or context, unlike {@link CameraPlugin}.
+   */
   public static void registerWith(Registrar registrar) {
-    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
+    if (registrar.activity() == null || Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
```

Comment:
We should talk about this. We recommend against this pattern because background vs foreground views are not distinguishable in most cases (a background view can switch to foreground).

Reply:
Got it, I'll remove this and update the below comment. I saw the comment and noticed that the code didn't match it and assumed that there was a mistake. The code must have been patched at some point and the comment wasn't updated.

Reply:
Done.

Reply:
I thought about this a bit more and we also want to keep the logic the same on both the old and new plugin registration path. So I moved this guard into `startListening` and renamed that to `maybeStartListening`.

/cc @cyanglaz this means that now sometimes it is expected that when we call `onDetached` `methodCallHandler` is null after all. So I removed the `Log.wtf` from the guard clause there and replaced it with a comment explaining the known case where that could happen.